### PR TITLE
Preliminary multi-window support for Obsidian 0.15

### PR DESCRIPTION
**Pull request by pjeby**
This is a draft of multi-window support for Obsidian 0.15.1+.  It currently works (mostly) except for the interactjs parts, and I suspect some of the bounding box calculations need to be fixed for secondary windows as well.

Basically, with this PR you can get hover editors in all windows, but you can't move/resize/header-double-tap the ones in secondary windows.  Unlike in the main window, the cursor also doesn't change to show move/resize options.

I tried adding code to ensure interact.addDocument() is called for every window's document (and removeDocument() on shutdown), but that did not fix the issue.  I may see if I can figure something else out later, but if you have any ideas I'd love to hear them.

### CODE REVIEW

**Comment by nothingislost** [comment]
Thanks pjeby! I'll try and look at this shortly.

**Comment by pjeby** [comment]
I got it working!  Had to create a separate interactStatic per window, and copy the default plugins to it.  There might be a cleaner way, but I dunno what it is.  I also haven't looked at the clipping calculations yet, but hover editors are now interactable in all windows with this version.  You probably have a better idea of what it's supposed to do, so have a look-see.

I'm pretty sure that one thing that's wrong right now is the calculation of the top -- I think it's subtracting the title bar size even though the area for secondary windows doesn't include the title bar in the first place.  There might be other issues.

**Comment by pjeby** [comment]
After a bit more playing, I've identified two things that are broken in (unmaximized) secondary windows:

1. Resizing upward sometimes allows you to place the header slightly off-screen, or sometimes doesn't show the resize until you release
2. Snapping to the right doesn't work, even though snapping left does

I think, though I'm not sure, that this might actually be an issue with resizing?

I've checked in a change to do reflows when secondary windows resize, but it doesn't fix snapping to the right or sizing to the top when a secondary window isn't maximized.  A second change to fix getting the body size info from the wrong document, and that seems to have fixed right-side snapping.

So, basically, sometimes dragging the top to resize in a secondary window still acts funny, not sure why.  Especially since it doesn't happen *every* time.

**Comment by pjeby** [comment]
One other weird issue: sometimes, if you hover a target in a secondary window, then move the mouse into the popover, leave it there for a few seconds and move it some more, the popover will disappear.  It's especially noticeable when the window is small enough that the popover overlaps the original target rectangle: in that case, just moving outside the target rectangle (while still over the poover) will close it.

I have no clue what to do with that one short of double-checking there aren't any more hidden references to the main document or window such that there are handlers being set up differently between the main and secondary windows.

**Comment by pjeby** [comment]
Ok, I think this is done for now.  I've scoured the code and they're probably some dumb little thing, but given that both problems (top-resize and sometimes-early-disappearance) are intermittent and *only* happen in secondary windows, this is probably good enough to use as a beta for multi-window support, and shouldn't break for people on the standard Obsidian version.

**Comment by nothingislost** [comment]
Thank you for all of the work you did on this!

I'm trying to test this out and am seeing issues with "Open in new window" on a leaf spawning empty windows rather than loading the document from the target leaf. The exception thrown was `n.matchParent` is not a function.

On doing a hard reload, it's preventing app startup do to `w.document.body.findAll` is not a function:

image

I'm still looking into whether this is some weird interaction on my end but just seeing if you have any ideas
